### PR TITLE
Add ABB-Welcome / Busch-Welcome Door Intercom Protocol

### PR DESCRIPTION
**Pull request by Mat931**
# What does this implement/fix?

This protocol allows monitoring your doorbell, opening your front door etc. with your ABB-Welcome / Busch-Welcome door intercom system. It requires a custom circuit to send and receive messages on the two-wire bus.

[Transmitter documentation](https://deploy-preview-2844--esphome.netlify.app/components/remote_transmitter.html#remote-transmitter-transmit-abbwelcome-action)
[Receiver documentation](https://deploy-preview-2844--esphome.netlify.app/components/remote_receiver.html?highlight=abbwelcome)

Requires https://github.com/esphome/esphome/pull/4642

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2844

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
esp32:
  board: esp32doit-devkit-v1
  framework:
    type: arduino

esphome:
  name: abb-welcome-demo
  friendly_name: "ABB Welcome Demo"
  on_boot:
    - lock.template.publish:
        id: main_door
        state: LOCKED
    - binary_sensor.template.publish:
        id: doorbell_indoor
        state: OFF
    - binary_sensor.template.publish:
        id: doorbell_outdoor
        state: OFF

wifi:
  networks:
  - ssid: !secret wifi_ssid
    password: !secret wifi_password

logger:

api:
  encryption:
    key: !secret api_encryption_key

ota:
  password: !secret ota_password

remote_transmitter:
  pin: GPIO26
  carrier_duty_percent: 100%

remote_receiver:
  pin:
    number: GPIO25
    mode:
      input: True
  dump: [abbwelcome, raw]
  filter: 8us
  tolerance: 26us # requires https://github.com/esphome/esphome/pull/4642
  idle: 1500us
  buffer_size: 15kB
  memory_blocks: 6
  clock_divider: 160 # requires https://github.com/esphome/esphome/pull/4642
  on_abbwelcome:
    then:
      - if:
          condition:
            and:
              - lambda: 'return (x.get_message_type() == 0x8d);' # unlock door response
              - lambda: 'return (x.get_source_address() == 0x4001);' # door address
          then:
            - lock.template.publish:
                id: main_door
                state: UNLOCKING
            - delay: 3s
            - lock.template.publish:
                id: main_door
                state: LOCKED
      - if:
          condition:
            and:
              - lambda: 'return (x.get_message_type() == 0x11);' # doorbell indoor
              - lambda: 'return (x.get_source_address() == 0x1001);' # your indoor station address
          then:
            - binary_sensor.template.publish:
                id: doorbell_indoor
                state: ON
            - binary_sensor.template.publish:
                id: doorbell_indoor
                state: OFF
      - if:
          condition:
            and:
              - lambda: 'return (x.get_message_type() == 0x01);' # doorbell outdoor
              - lambda: 'return (x.get_source_address() == 0x2001);' # outdoor station address
              - lambda: 'return (x.get_destination_address() == 0x1001);' # your indoor station address
          then:
            - binary_sensor.template.publish:
                id: doorbell_outdoor
                state: ON
            - binary_sensor.template.publish:
                id: doorbell_outdoor
                state: OFF

binary_sensor:
  - platform: template
    name: Doorbell indoor
    id: doorbell_indoor
    icon: mdi:bell
    filters:
      - delayed_off: 1s

  - platform: template
    name: Doorbell outdoor
    id: doorbell_outdoor
    icon: mdi:bell
    filters:
      - delayed_off: 1s

lock:
  - platform: template
    name: "Main Door"
    id: main_door
    on_unlock:
      - then:
          - lock.template.publish:
              id: main_door
              state: LOCKED
    lock_action:
      - then:
          - lock.template.publish:
              id: main_door
              state: LOCKED
    unlock_action:
      - then:
          - remote_transmitter.transmit_abbwelcome:
              source_address: 0x1001 # your indoor station address
              destination_address: 0x4001 # door address
              message_type: 0x0d # unlock door
              data: [0xab, 0xcd, 0xef]  # door opener secret code, see receiver dump

button:
  - platform: template
    name: "Stop Doorbell"
    on_press:
      - remote_transmitter.transmit_abbwelcome:
          source_address: 0x1001 # your indoor station address
          destination_address: 0x2001 # outdoor station address
          message_type: 0x02 # end call (stops your doorbell ringtone)
          data: [0x00]
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by Mat931** [comment]
<details>
<summary>Circuit</summary>

circuit
</details>

**Comment by dylanmazurek** [comment]
I love this functionality! I've ordered all the parts to create the circuit and will be adding it to my intercom as soon as I build it. Question about the implementation, does it work in line with the intercom screen or does it replace the screen?

**Comment by Mat931** [comment]
@DylanMazurek You can add it in parallel with the intercom screen or replace it.

**Comment by bilak** [comment]
Hey, I'm doing research how to make my ABB welcome little bit smarter and found this. Is there any chance that this will become a part of esphome? I didn't touched the esphome yet but if this is something that should work I'd give it a try. 

Dumb question: is there any possibility to include also the sound when somebody speaks or even do some 2 way audio so it would be possible to create some integration with sound?

**Comment by Mat931** [comment]
@bilak Not sure if there are any plans to merge this into ESPHome at this point. You can still use it as an external component. With the circuit above you can only receive and send control messages, no audio or video. You can find more info in the comments [here](https://github.com/esphome/esphome-docs/pull/2844). For me personally this project is finished because I have all the functionality I need and I already ordered my custom PCBs with the circuit years ago. But feel free to investigate more. Good Luck!

**Comment by dylanmazurek** [comment]
Hey @Mat931 I'm looking into this again, is there any chance you have the circuit files you sent to get made? I haven't really worked with circuit design or fab so it would be awesome if you had yours lying around 😊

**Comment by Mat931** [comment]
@dylanmazurek Sure I still have the PCB files. Before I share them I want to clean up some errors and remove unused footprints. The PCB can be installed on top of a flush-mounted box on a wall. It has some touch buttons and LEDs so you can program it as a standalone door opener for example. The board can be powered from the ABB-Welcome bus and send/receive control messages. (no audio or video) You would still need to solder the components to the PCB. Most are 0603 and bigger but there are also some smaller ones. Do you think that works for you?
<details>
  <summary>PCB images</summary>

  abb-welcome-esphome

  abb-welcome-esphome-back

</details>

**Comment by dylanmazurek** [comment]
Wow that would be amazing! I think what I would do is send the files and components required to a company that can do it for me (my soldering skills are very much lacking), I found one that's around the corner from me and hopefully they can help me out.

**Comment by Mat931** [comment]
@dylanmazurek I designed a new board that only has the minimum of required components to supply an ESP32 with power and send/receive messages. In theory you can order 2 of the fully assembled boards for under 50€ from JLC. (I haven't tried)
If you have Discord you can contact me there @mat931 and we can discuss the details.
abb-welcome-esphome-simplified

File: esphome/components/remote_base/abbwelcome_protocol.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    }
+  }
+  next_space += BYTE_SPACE_US;
+  dst->space(next_space);
+}
+
+void ABBWelcomeProtocol::encode(RemoteTransmitData *dst, const ABBWelcomeData &src) {
+  dst->set_carrier_frequency(0);
+  uint32_t reserve_count = 0;
+  for (unsigned i = 0; i < src.size(); i++) {
```
Comment (jesserockz) [inline esphome/components/remote_base/abbwelcome_protocol.cpp]:
```suggestion
  for (size_t i = 0; i < src.size(); i++) {
```

File: esphome/components/remote_base/abbwelcome_protocol.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    reserve_count += 2 * (9 - (src[i] & 1) - ((src[i] >> 1) & 1) - ((src[i] >> 2) & 1) - ((src[i] >> 3) & 1) -
+                          ((src[i] >> 4) & 1) - ((src[i] >> 5) & 1) - ((src[i] >> 6) & 1) - ((src[i] >> 7) & 1));
+  }
+  dst->reserve(reserve_count);
+  for (unsigned i = 0; i < src.size(); i++)
```
Comment (jesserockz) [inline esphome/components/remote_base/abbwelcome_protocol.cpp]:
```suggestion
  for (size_t i = 0; i < src.size(); i++)
```

File: esphome/components/remote_base/abbwelcome_protocol.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    encode_byte_(dst, src[i]);
```
Comment (jesserockz) [inline esphome/components/remote_base/abbwelcome_protocol.cpp]: Please make sure to prefix all class members and member functions with `this->`
```suggestion
    this->encode_byte_(dst, src[i]);
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/remote_base/abbwelcome_protocol.cpp (hunk 3 above)
Comment (Mat931) [reply esphome/components/remote_base/abbwelcome_protocol.cpp]: Thanks for the review! I try to add `this->` to all class members but seems I missed a few places. Do you know if there is an automated method to find the places where `this->` is missing?

Comment (jesserockz) [reply esphome/components/remote_base/abbwelcome_protocol.cpp]: I don't know of any automated way sorry.

File: esphome/components/remote_base/abbwelcome_protocol.h [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    uint8_t data_length : 4;
+    uint8_t reply : 1;
+    uint8_t message_type : 7;
+    uint16_t destination_address;
+    uint16_t source_address;
+    uint8_t message_id;
+    uint8_t data[8];
+  };
+};
+*/
```
Comment (jesserockz) [inline esphome/components/remote_base/abbwelcome_protocol.h]: No commented out code please
```suggestion
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**